File: src/test/java/no/fintlabs/access/control/service/ResourceAccessServiceTest.java
package no.fintlabs.access.control.service;

import no.fintlabs.access.control.model.dto.ReadingOption;
import no.fintlabs.access.control.model.dto.ResourceAccess;
import no.fintlabs.access.control.model.entity.ClientEntity;
import no.fintlabs.access.control.model.entity.PackageAccessEntity;
import no.fintlabs.access.control.model.entity.ResourceAccessEntity;
import no.fintlabs.access.control.model.metamodel.Metamodel;
import no.fintlabs.access.control.model.metamodel.MetamodelRepository;
import no.fintlabs.access.control.repository.ClientEntityRepository;
import no.fintlabs.access.control.repository.PackageAccessEntityRepository;
import no.fintlabs.access.control.repository.ResourceAccessEntityRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.*;

class ResourceAccessServiceTest {

    @Mock
    private ClientEntityRepository clientEntityRepository;

    @Mock
    private ResourceAccessEntityRepository resourceAccessEntityRepository;

    @Mock
    private PackageAccessEntityRepository packageAccessEntityRepository;

    @Mock
    private MetamodelRepository metamodelRepository;

    @Mock
    private FieldAccessService fieldAccessService;

    private ResourceAccessService resourceAccessService;

    @BeforeEach
    void setUp() {
        MockitoAnnotations.openMocks(this);
        resourceAccessService = new ResourceAccessService(
                clientEntityRepository,
                resourceAccessEntityRepository,
                packageAccessEntityRepository,
                metamodelRepository,
                fieldAccessService
        );
    }

    @Test
    void getResourceAccessWithNoClient() {
        String clientName = "test-client";
        String component = "test-component";

        when(clientEntityRepository.findById(clientName)).thenReturn(Optional.empty());

        Collection<ResourceAccess> result = resourceAccessService.getResourceAccess(clientName, component);

        assertEquals(0, result.size());
        verify(clientEntityRepository).findById(clientName);
        verify(metamodelRepository).getResourceAccessByComponent(component);
        verifyNoInteractions(resourceAccessEntityRepository, packageAccessEntityRepository, fieldAccessService);
    }

    @Test
    void getResourceAccessWithClientAndNoPackage() {
        String clientName = "test-client";
        String component = "test-component";
        ClientEntity clientEntity = new ClientEntity();

        when(clientEntityRepository.findById(clientName)).thenReturn(Optional.of(clientEntity));
        when(packageAccessEntityRepository.findByClient(clientEntity)).thenReturn(Collections.emptyList());

        Collection<ResourceAccess> result = resourceAccessService.getResourceAccess(clientName, component);

        assertEquals(0, result.size());
        verify(metamodelRepository).getResourceAccessByComponent(component);
        verify(clientEntityRepository).findById(clientName);
        verify(packageAccessEntityRepository).findByClient(clientEntity);
        verifyNoInteractions(resourceAccessEntityRepository, fieldAccessService);
    }

    @Test
    void getResourceAccessWithValidData() {
        String clientName = "test-client";
        String component = "test-component";
        ClientEntity clientEntity = new ClientEntity();

        PackageAccessEntity packageAccessEntity = new PackageAccessEntity();
        packageAccessEntity.setId(123);
        packageAccessEntity.setDomainName("test");
        packageAccessEntity.setPackageName("component");

        ResourceAccessEntity resourceAccessEntity = new ResourceAccessEntity();
        resourceAccessEntity.setId(321);
        resourceAccessEntity.setResourceName("elevforhold");

        Metamodel metamodel = new Metamodel(
                "test",
                "component",
                "elevforhold",
                List.of("field1", "field2"),
                true
        );

        ResourceAccess resourceAccess = new ResourceAccess("elevforhold", List.of(), ReadingOption.SINGULAR, false, false);

        when(clientEntityRepository.findById(clientName)).thenReturn(Optional.of(clientEntity));
        when(packageAccessEntityRepository.findByClient(clientEntity)).thenReturn(List.of(packageAccessEntity));
        when(resourceAccessEntityRepository.findByPackageId(packageAccessEntity.getId()))
                .thenReturn(List.of(resourceAccessEntity));
        when(metamodelRepository.getResourceAccessByComponent(component)).thenReturn(List.of(metamodel));
        when(fieldAccessService.getFieldAccess(metamodel, Optional.of(resourceAccessEntity))).thenReturn(resourceAccess.fields());

        Collection<ResourceAccess> result = resourceAccessService.getResourceAccess(clientName, component);

        assertEquals(1, result.size());
        ResourceAccess resultAccess = result.iterator().next();
        assertEquals("elevforhold", resultAccess.name());
        assertEquals(ReadingOption.SINGULAR, resultAccess.readingOption());

        verify(clientEntityRepository).findById(clientName);
        verify(packageAccessEntityRepository).findByClient(clientEntity);
        //verify(resourceAccessEntityRepository).findByPackageId(packageAccessEntity.getId());
        verify(metamodelRepository).getResourceAccessByComponent(component);
        verify(fieldAccessService).getFieldAccess(metamodel, Optional.of(resourceAccessEntity));
    }
}

File: src/test/kotlin/no/fintlabs/access/control/service/ResourceAccessServiceTest.kt
package no.fintlabs.access.control.service

import no.fintlabs.access.control.model.dto.ReadingOption
import no.fintlabs.access.control.model.dto.ResourceAccess
import no.fintlabs.access.control.model.entity.ClientEntity
import no.fintlabs.access.control.model.entity.PackageAccessEntity
import no.fintlabs.access.control.model.entity.ResourceAccessEntity
import no.fintlabs.access.control.model.metamodel.Metamodel
import no.fintlabs.access.control.model.metamodel.MetamodelRepository
import no.fintlabs.access.control.repository.ClientEntityRepository
import no.fintlabs.access.control.repository.PackageAccessEntityRepository
import no.fintlabs.access.control.repository.ResourceAccessEntityRepository
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.MockitoAnnotations
import java.util.*

internal class ResourceAccessServiceTest {

    @Mock
    lateinit var clientEntityRepository: ClientEntityRepository

    @Mock
    lateinit var resourceAccessEntityRepository: ResourceAccessEntityRepository

    @Mock
    lateinit var packageAccessEntityRepository: PackageAccessEntityRepository

    @Mock
    lateinit var metamodelRepository: MetamodelRepository

    @Mock
    lateinit var fieldAccessService: FieldAccessService

    lateinit var resourceAccessService: ResourceAccessService

    @BeforeEach
    fun setUp() {
        MockitoAnnotations.openMocks(this)
        resourceAccessService = ResourceAccessService(
            clientEntityRepository,
            resourceAccessEntityRepository,
            packageAccessEntityRepository,
            metamodelRepository,
            fieldAccessService
        )
    }

    @Test
    fun resourceAccessWithNoClient() {
        val clientName = "test-client"
        val component = "test-component"

        Mockito.`when`(clientEntityRepository.findById(clientName)).thenReturn(Optional.empty())

        val result = resourceAccessService.getResourceAccess(clientName, component)

        Assertions.assertEquals(0, result.size)
        Mockito.verify(clientEntityRepository).findById(clientName)
        Mockito.verify(metamodelRepository).getResourceAccessByComponent(component)
        Mockito.verifyNoInteractions(
            resourceAccessEntityRepository,
            packageAccessEntityRepository,
            fieldAccessService
        )
    }

    @Test
    fun resourceAccessWithClientAndNoPackage() {
        val clientName = "test-client"
        val component = "test-component"
        val clientEntity = ClientEntity()

        Mockito.`when`(clientEntityRepository.findById(clientName)).thenReturn(Optional.of(clientEntity))
        Mockito.`when`(packageAccessEntityRepository.findByClient(clientEntity)).thenReturn(emptyList())

        val result = resourceAccessService.getResourceAccess(clientName, component)

        Assertions.assertEquals(0, result.size)
        Mockito.verify(metamodelRepository).getResourceAccessByComponent(component)
        Mockito.verify(clientEntityRepository).findById(clientName)
        Mockito.verify(packageAccessEntityRepository).findByClient(clientEntity)
        Mockito.verifyNoInteractions(resourceAccessEntityRepository, fieldAccessService)
    }

    @Test
    fun resourceAccessWithValidData() {
        val clientName = "test-client"
        val component = "test-component"
        val clientEntity = ClientEntity()

        val packageAccessEntity = PackageAccessEntity(
            id = 123,
            domainName = "test",
            packageName = "component",
            client = ClientEntity()
        )

        val resourceAccessEntity = ResourceAccessEntity(id = 321, resourceName = "elevforhold")

        val metamodel =
            Metamodel(
                "test",
                "component",
                "elevforhold",
                listOf("field1", "field2"),
                true
            )

        val resourceAccess = ResourceAccess(
            "elevforhold",
            listOf(),
            ReadingOption.SINGULAR,
            false,
            false
        )

        Mockito.`when`(clientEntityRepository.findById(clientName)).thenReturn(Optional.of(clientEntity))
        Mockito.`when`(packageAccessEntityRepository.findByClient(clientEntity)).thenReturn(listOf(packageAccessEntity))
        Mockito.`when`(resourceAccessEntityRepository.findByPackageId(packageAccessEntity.id ?: 0))
            .thenReturn(listOf(resourceAccessEntity))
        Mockito.`when`(metamodelRepository.getResourceAccessByComponent(component)).thenReturn(listOf(metamodel))
        Mockito.`when`(fieldAccessService.getFieldAccess(metamodel, resourceAccessEntity))
            .thenReturn(resourceAccess.fields)

        val result = resourceAccessService.getResourceAccess(clientName, component)

        Assertions.assertEquals(1, result.size)
        val resultAccess = result.iterator().next()
        Assertions.assertEquals("elevforhold", resultAccess.name)
        Assertions.assertEquals(ReadingOption.SINGULAR, resultAccess.readingOption)

        Mockito.verify(clientEntityRepository).findById(clientName)
        Mockito.verify(packageAccessEntityRepository).findByClient(clientEntity)
        //verify(resourceAccessEntityRepository).findByPackageId(packageAccessEntity.getId());
        Mockito.verify(metamodelRepository).getResourceAccessByComponent(component)
        Mockito.verify(fieldAccessService).getFieldAccess(metamodel, resourceAccessEntity)
    }
}
